Implement #543: allow custom typescript compiler (ttypescript, ntypescript, etc)

cspotcode opened the pull request:
If ts-jest configuration has a `compiler` property, it is `require()`d instead of `typescript` to use as the compiler.  There are a few alternative compilers out there that support extra features.  ntypescript or ttypescript.

GeeWee commented on src/custom-compiler.ts:
```diff
@@ -0,0 +1,15 @@
+import * as __typeof__ts from 'typescript';
```
this seems like a really strange name, also given the typeof __typeof_ts statement below - can we make this clearer somehow?

cspotcode replied:
I like your `ts_types` suggestion.

GeeWee commented on src/custom-compiler.ts:
```diff
@@ -0,0 +1,15 @@
+ */
+export function getTypescriptCompiler(config: TsJestConfig): typeof __typeof__ts {
+  if (typeof config.compiler === 'string') {
+    return require(config.compiler);
```
Perhaps we should have a catch block for these dynamic requires? Just to give a better error message if it goes wrong.

cspotcode replied:
Any ideas what it should say?  Node's default error is:
```
Error: Cannot find module 'ttypescript'
    at Function.Module._resolveFilename (module.js:542:15)
```

Should we wrap it with this prefix?
```
Error: Importing TypeScript compiler: Cannot find module 'ttypescript'
```

GeeWee commented on src/transpiler.ts:
```diff
@@ -1,22 +1,25 @@
 import { cwd } from 'process';
-import * as ts from 'typescript';
+// Only for types; take care never to use ts_ in expressions, only in type annotations
+import * as ts_ from 'typescript';
```
let's call it ts_types or something a little more meaningful

GeeWee replied:
Or instead of using typeof everywhere - perhaps export a type that's defined as

cspotcode replied:
Unfortunately type aliases cannot be used with namespaces, so `type TsTypes = typeof ts´ is rejected by the compiler.  I'll use `ts_types`.

cspotcode commented:
I've asked over in #543 what's the best way to test with `ttypescript`.  Then I'll push a unit test.

cspotcode commented:
@kulshekhar It's impossible to use ts-jest on projects that compile with custom transformers.  ts-jest could add built-in support for custom transformers, but after reading ttypescript's documentation, I think it's better to allow using ttypescript.  It handles a bunch of issues automatically and the API is identical to typescript, so ts-jest still calls `transpileModule` like normal.

I rebased changes to master and added a test case that also demonstrates why someone might want to use `ttypescript` instead of `typescript`.

cspotcode commented:
@GeeWee @kulshekhar Tests are passing; let me know if you have any questions, and thanks in advance.

kulshekhar commented:
@cspotcode I'm so sorry for the delay on this. I've been hoping to find some time to review this but haven't been able to. If I can't find the time by this weekend, I'll just merge this (and some other PRs) in - would that be fine?

cspotcode commented:
@kulshekhar no rush at all; the delay is my fault.  I finally got the unit test working yesterday.

Regarding this commit:
`Fix bug in getTSConfig where it omits rootDir from cache key`

It is technically an unrelated bug.  I can move it into a separate PR or remove it depending on what you think.

kulshekhar commented:
@cspotcode I've merged the other PRs in. If the conflict in this PR can be resolved, I'll merge it in as well and include the changes in a new release

cspotcode commented:
@kulshekhar I merged the latest master, resolved all conflicts, and fixed failing tests.  This should be ready to merge.  Thank you!
